### PR TITLE
add ability to exclude rules from namespace label enforcement

### PR DESCRIPTION
**Pull request by sev3ryn**
fixes https://github.com/coreos/prometheus-operator/issues/2940

### CODE REVIEW

File: Documentation/api.md [hunk 1]
```diff
@@ -314,6 +315,17 @@ PodMonitorSpec contains specification parameters for a PodMonitor.
 
 [Back to TOC](#table-of-contents)
 
+## PromRuleExcludeConfig
```
Comment (brancz) [inline Documentation/api.md:318]: We tend to not abbreviate, especially for types.

**Comment by sev3ryn** [comment]
fixed test cases and added tests for new functionality

**Comment by sev3ryn** [comment]
I didn't noticed that changed functions are also used for Thanos rules - will fix it as soon as I can

**Comment by brancz** [comment]
Actually now that you say it, we're going to have to add the same functionality to the Thanos ruler.

**Comment by sev3ryn** [comment]
added functionality to ThanosRuler. 
I adhered style of operator - by this I mean almost "copy-paste" pieces of code between prometheus/rules and thanos/rule. Shouldn't I move nsLabelEnforcementExcludeList and its methods to separate package to follow DRY principle?

If not, could you please elaborate why - just to educate myself.

**Comment by brancz** [comment]
> Shouldn't I move nsLabelEnforcementExcludeList and its methods to separate package to follow DRY principle?

Please feel free to do so, we wanted to do so anyways.

**Comment by sev3ryn** [comment]
moved namespace label enforcement to shared library. Naming suggestions are welcome

**Comment by sev3ryn** [comment]
@brancz ^^

**Comment by brancz** [comment]
sorry looks like this needs a rebase

**Comment by sev3ryn** [comment]
rebased

File: pkg/apis/monitoring/v1/types.go [hunk 2]
```diff
@@ -315,6 +319,15 @@ type PrometheusSpec struct {
 	QueryLogFile string `json:"queryLogFile,omitempty"`
 }
 
+// PrometheusRuleExcludeConfig enables users to configure excluded PrometheusRule names and their namespaces
+// to be ignored while enforcing namespace label for alerts and metrics.
+type PrometheusRuleExcludeConfig struct {
+	// RuleNamespace - namespace of excluded rule
+	RuleNamespace string `json:"ruleNamespace,omitempty"`
```
Comment (brancz) [inline pkg/apis/monitoring/v1/types.go:326]: this shouldn't be `omitempty` I think, as that marks the field as not required, but if an object of this type is specified then the namespace and name are mandatory, just the list itself is optional

File: pkg/namespace-labeler/labeler.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	excludeList     map[string]map[string]struct{}
+}
+
+// New - creates new Labeler
+// enforcedNsLabel - label name to be enforced for namespace
+// excludeConfig - list of namespace + PrometheusRule names to be excluded while enforcing adding namespace label
+func New(enforcedNsLabel string, excludeConfig []monitoringv1.PrometheusRuleExcludeConfig) *Labeler {
+
+	if enforcedNsLabel == "" || len(excludeConfig) == 0 {
+		return &Labeler{enforcedNsLabel: enforcedNsLabel}
```
Comment (brancz) [inline pkg/namespace-labeler/labeler.go:39]: can we add a comment here that this is essentially a no-op labeler? Just for me in 3months to understand what this is :)

File: pkg/prometheus/rules.go [hunk 4]
```diff
@@ -208,34 +216,8 @@ func (c *Operator) selectRules(p *monitoringv1.Prometheus, namespaces []string)
...
-					return "", errors.Wrap(err, "failed to inject labels to expression")
-				}
-
-				promRule.Groups[gi].Rules[ri].Expr = intstr.FromString(parsedExpr.String())
-			}
-		}
-	}
+func generateContent(promRule monitoringv1.PrometheusRuleSpec) (string, error) {
+
 	content, err := yaml.Marshal(promRule)
```
Comment (brancz) [inline pkg/prometheus/rules.go:221]: I don't see us doing this previously, so I'm fine with this being a follow up, but we need to discard the `partial_response_strategy` (also shouldn't be snake case).

Comment (sev3ryn) [reply pkg/prometheus/rules.go:221]: you do - see [pkg/prometheus/rules.go#L214](https://github.com/coreos/prometheus-operator/blob/master/pkg/prometheus/rules.go#L214)

snake case is an issue of Thanos unfortunately. See [partial-response](https://github.com/thanos-io/thanos/blob/v0.4.0-rc.0/docs/components/rule.md#partial-response)

but it is an issue - I'm discarding this field for Thanos Rules also what I shouldn't do - will fix it

**Comment by brancz** [comment]
Awesome! Thank you so much for this!

**Pull request merged**